language-support/ts/daml-ledger/index.ts

```diff
@@ -616,19 +666,58 @@ class Ledger {
    * @typeparam T The contract template type.
    * @typeparam K The contract key type.
    * @typeparam I The contract id type.
+   *
+   * @deprecated since 1.5; use [[streamFetchByKeys]] instead
    */
+  // This code is deprecated and only kept to guarantee backwardss
+  // compatibility. Do not change until removal.
   streamFetchByKey<T extends object, K, I extends string>(
     template: Template<T, K, I>,
     key: K,
   ): Stream<T, K, I, CreateEvent<T, K, I> | null> {
     let lastContractId: ContractId<T> | null = null;
     const request = [{templateId: template.templateId, key}];
     const reconnectRequest = (): object[] => [{...request[0], 'contractIdAtOffset': lastContractId}]
+    const change = (contract: CreateEvent<T, K, I> | null, events: readonly Event<T, K, I>[]): CreateEvent<T, K, I> | null => {
+      for (const event of events) {
+        if ('created' in event) {
+          contract = event.created;
+        } else {
+          if (contract && contract.contractId === event.archived.contractId) {
+            contract = null;
+          }
+        }
+      }
+      lastContractId = contract ? contract.contractId : null
+      return contract;
+    }
+    return this.streamSubmit(template, 'v1/stream/fetch', request, reconnectRequest, null, change);
+  }
+
+  /**
+   * Retrieve a consolidated stream of events for a given template and contract key.
+   *
+   * Same as [[streamQuery]], but instead of a query, match contracts by
+   * contract key. If multiple keys are given, returns a consolidated stream of
+   * all events for all given keys.
+   *
+   * @typeparam T The contract template type.
+   * @typeparam K The contract key type.
+   * @typeparam I The contract id type.
+   */
+  streamFetchByKeys<T extends object, K, I extends string>(
+    template: Template<T, K, I>,
+    key: K,
+    ...keys: K[]
+  ): Stream<T, K, I, CreateEvent<T, K, I> | null> {
```

Comment:
I don’t think this implementation works. You are only tracking a single contract key but you are querying for multiple contract keys. Here there is actually a meaningful difference between the singular and the plural version:
1. For the singular version you are tracking the state of of the unique contract currently associated with the key.
2. For the plural version you are tracking the evolution of multiple contracts. I see two possible APIs here (but there are definitely more):
    1. The API provided by `useStreamQueries`. This basically means that you’re ignoring uniqueness of contract keys in the API and just use it as a special kind of query. Users could of course still track the current contract associated with each key.
    2. Stream an array of `CreatedEvent<T,K,I>| null` of length `n` where `n` is the number of keys you specified.

I feel like `2` is probably what people want. Consider, a UI where you want to track two contract keys you would just end up building 2 on top of 1 if the API does not provide you with it.

Reply:
You're right. When first opening this PR, I was too focused on whether deprecation & single-template were the right approaches and did not look at the actual behaviour enough.

I agree 2.ii is also what I would expect here; I'll change it to match that.